File: core/src/main/java/com/hxw/core/utils/AppUtils.java
package com.hxw.lol.utils;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.Size;
import android.support.design.widget.Snackbar;
import android.support.v4.content.ContextCompat;
import android.view.View;
import android.widget.Toast;

import com.hxw.lol.delegate.AppDelegate;

/**
 * App相关工具类
 *
 * @author hxw on 2018/5/5.
 */
public class AppUtils {
    private static Toast mToast;
    public static void showSnackBar(String message) {
        Activity currentActivity = AppDelegate
                .getAppComponent()
                .appManager()
                .getCurrentActivity();
        if (currentActivity != null) {
            View view = currentActivity
                    .getWindow().getDecorView().findViewById(android.R.id.content);
            Snackbar.make(view, message, Snackbar.LENGTH_SHORT).show();
        }
    }

    @SuppressLint("ShowToast")
    public static void showToast(String message){
        if (mToast==null){
            mToast=Toast.makeText(AppDelegate.getAppComponent().application(),
                    "",Toast.LENGTH_SHORT);
        }
        mToast.setText(message);
        mToast.show();
    }
    public static int getVersionCode(Context context) {
        try {
            PackageInfo info = context.getPackageManager()
                    .getPackageInfo(context.getPackageName(), 0);
            return info.versionCode;
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
            return 0;
        }
    }

    public static String getVersionName(Context context) {
        try {
            PackageInfo info = context.getPackageManager()
                    .getPackageInfo(context.getPackageName(), 0);
            return info.versionName;
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
            return "";
        }
    }

    /**
     * 检查权限
     * 有权限: PackageManager.PERMISSION_GRANTED
     * 无权限: PackageManager.PERMISSION_DENIED
     *
     * @param context 上下文
     * @param perms   权限
     * @return 是否有权限 {@code true} 有权限 {@code false} 无权限
     */
    public static boolean hasPermissions(@NonNull Context context,
                                         @Size(min = 1) @NonNull String... perms) {
        //版本6.0以下不需要请求权限
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            return true;
        }
        //是否有权限
        for (String perm : perms) {
            if (ContextCompat.checkSelfPermission(context, perm)
                    != PackageManager.PERMISSION_GRANTED) {
                return false;
            }
        }
        return true;
    }

    /**
     * dp转px
     *
     * @param dpValue dp值
     * @return px值
     */
    public static int dpToPx(Context context,float dpValue){
        float scale= context.getResources().getDisplayMetrics().density;
        return (int) (dpValue * scale + 0.5f);
    }

    /**
     * sp转px
     *
     * @param spValue sp值
     * @return px值
     */
    public static int spToPx(Context context,float spValue){
        float scale= context.getResources().getDisplayMetrics().scaledDensity;
        return (int) (spValue * scale + 0.5f);
    }
}

File: core/src/main/java/com/hxw/core/utils/AppUtils.kt
package com.hxw.core.utils

import android.content.Context
import android.content.pm.PackageManager
import android.net.ParseException
import android.os.Build
import android.view.View
import com.google.gson.JsonParseException
import com.hxw.core.integration.AppManager
import org.jetbrains.anko.design.snackbar
import org.jetbrains.anko.toast
import org.json.JSONException
import retrofit2.HttpException
import timber.log.Timber
import java.net.NoRouteToHostException
import java.net.SocketTimeoutException
import java.net.UnknownHostException


/**
 * App相关工具类
 *
 * @author hxw
 * @date 2018/5/5
 */


/**
 * 通过SnackBar展示信息
 *
 * @param message 信息
 */

fun showSnackBar(message: String) {
    val topActivity = AppManager.getTopActivity()
    if (topActivity != null) {
        val view = topActivity
            .window.decorView.findViewById<View>(android.R.id.content)
        view.snackbar(message)
    }
}

/**
 * 通过Toast展示信息
 *
 * @param message 信息
 */

fun showToast(message: String) {
    val topActivity = AppManager.getTopActivity()
    topActivity?.toast(message)
}

/**
 * 获取版本号
 */

fun Context.getVersionCode(): Long {
    return try {
        val info = this.packageManager
            .getPackageInfo(this.packageName, 0)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            info.longVersionCode
        } else {
            info.versionCode.toLong()
        }
    } catch (e: PackageManager.NameNotFoundException) {
        e.printStackTrace()
        0
    }

}

/**
 * 获取版本名
 */

fun Context.getVersionName(): String {
    return try {
        val info = this.packageManager
            .getPackageInfo(this.packageName, 0)
        info.versionName
    } catch (e: PackageManager.NameNotFoundException) {
        e.printStackTrace()
        ""
    }
}

/**
 * 错误处理
 */

fun Throwable?.onError() {
    Timber.tag("Catch-Error").e(this)
    val msg = when (this) {
        null -> "无报错信息Throwable==null"
        is UnknownHostException -> "域名解析失败,请检查网络和服务器"
        is SocketTimeoutException -> "请求网络超时"
        is NoRouteToHostException -> "无法连接远程地址与端口"
        is HttpException -> when (this.code()) {
            500 -> "服务器发生错误"
            404 -> "请求地址不存在"
            403 -> "请求被服务器拒绝"
            307 -> "请求被重定向到其他页面"
            else -> this.message()
        }
        is JsonParseException, is ParseException, is JSONException -> "数据解析错误"
        else -> this.message.toString()
    }
    showSnackBar(msg)
}

